core/modules/loader/ServerTcpBase.cc

```diff
@@ -0,0 +1,618 @@
+        LOGS(_log, LOG_LVL_INFO, funcName <<  " parsing bytes=" << bytesTrans <<
+                                 " _buf" << _buf.dumpStr(false));
+        auto protoItem = StringElement::protoParse<proto::WorkerKeysInfo>(_buf);
+        if (protoItem == nullptr) {
+            throw LoaderMsgErr(ERR_LOC, "protoItem nullptr");
+        }
+        NeighborsInfo nInfo;
+        auto workerName = protoItem->wid();
+        nInfo.keyCount = protoItem->mapsize();
+        nInfo.recentAdds = protoItem->recentadds();
+        proto::WorkerRangeString protoRange = protoItem->range();
+        LOGS(_log, LOG_LVL_INFO, funcName << " WorkerKeysInfo name=" << workerName <<
+                                 " keyCount=" << nInfo.keyCount << " recentAdds=" << nInfo.recentAdds);
+        bool valid = protoRange.valid();
+        StringRange leftRange;
+        StringRange newLeftRange;
+        if (valid) {
+            std::string min   = protoRange.min();
+            std::string max   = protoRange.max();
+            bool unlimited = protoRange.maxunlimited();
+            leftRange.setMinMax(min, max, unlimited);
+            LOGS(_log, LOG_LVL_WARN, funcName << " leftRange=" << leftRange);
+            newLeftRange = _serverTcpBase->getCentralWorker()->updateRangeWithLeftData(leftRange);
+        }
+        proto::Neighbor protoLeftNeigh = protoItem->left();
+        nInfo.neighborLeft->update(protoLeftNeigh.wid());  // Not really useful in this case.
+        proto::Neighbor protoRightNeigh = protoItem->right();
+        nInfo.neighborRight->update(protoRightNeigh.wid()); // This should be our name
+        if (nInfo.neighborRight->get() != _serverTcpBase->getOurName()) {
+            LOGS(_log, LOG_LVL_ERROR, "Our (" << _serverTcpBase->getOurName() <<
+                                      ") left neighbor does not have our name as its right neighbor" );
+        }
+
+        _serverTcpBase->getCentralWorker()->setNeighborInfoLeft(workerName, nInfo.keyCount, newLeftRange);
+
+        // Need to send our range and key count back to left neighbor so it can figure out what to do with its range.
+        _buf.reset();
+        StringElement strWKI;
+        std::unique_ptr<proto::WorkerKeysInfo> protoWKI = _serverTcpBase->getCentralWorker()->_workerKeysInfoBuilder();
+        protoWKI->SerializeToString(&(strWKI.element));
+        UInt32Element bytesInMsg(strWKI.transmitSize());
+        // Send the number of bytes in the message so TCP client knows how many bytes to read.
+        bytesInMsg.appendToData(_buf);
+        strWKI.appendToData(_buf);
+        ServerTcpBase::writeData(_socket, _buf);
+        LOGS(_log, LOG_LVL_INFO, funcName << " done");
+    } catch (LoaderMsgErr &msgErr) {
+        LOGS(_log, LOG_LVL_ERROR, funcName << " Buffer failed " << msgErr.what());
+        _freeConnect();
+        return;
+    }
+    boost::system::error_code ecode;
+    _readKind(ecode, 0); // get next message
```

Comment:
Are you explicitly calling the same callback method which is normally called asynchronously?
May be there is a way to refactor the code of into the async handler and a method which would be getting the next message? Then you would be able to call that second method directly from here w/o making a complicated (and possibly fragile) implementation of `_readKind` which would be called from anywhere?

Reply:
@jgates108 It would be helpful for me coming through as 2nd reviewer if there were answers or comments on these questions.

Reply:
I am calling _readKind to move on to the next element. I'm open to suggestions for a cleaner way to do it.